packages/botonic-plugin-flow-builder/src/api.ts

```diff
@@ -0,0 +1,163 @@
+    ) as HtStartNode | undefined
+    if (!startUpNode) throw new Error('Start-up id must be defined')
+    return this.getNodeById(startUpNode.target.id)
+  }
+
+  getFallbackNode(alternate: boolean): HtNodeWithContent {
+    const fallbackNode = this.flow.nodes.find(
+      node => node.type === HtNodeWithContentType.FALLBACK
+    ) as HtFallbackNode | undefined
+    if (!fallbackNode) {
+      throw new Error('Fallback node must be defined')
+    }
+    const fallbackFirstMessage = fallbackNode.content.first_message
+    if (!fallbackFirstMessage) {
+      throw new Error('Fallback 1st message must be defined')
+    }
+    const fallbackSecondMessage = fallbackNode.content.second_message
+    if (!fallbackSecondMessage) {
+      return this.getNodeById(fallbackFirstMessage.id)
+    }
+    return alternate
+      ? this.getNodeById(fallbackFirstMessage.id)
+      : this.getNodeById(fallbackSecondMessage.id)
+  }
+
+  getNodeByIntent(input: Input, locale: string): HtNodeWithContent | undefined {
+    try {
+      const intents = this.flow.nodes.filter(
+        node => node.type === HtNodeWithContentType.INTENT
+      ) as HtIntentNode[]
+      const inputIntent = input.intent
+      const inputConfidence = input.confidence
+      if (inputIntent) {
+        const matchedIntentNode = intents.find(
+          node =>
+            inputIntent &&
+            this.hasIntent(node, inputIntent, locale) &&
+            inputConfidence &&
+            this.hasMetConfidenceThreshold(node, inputConfidence)
+        )
+        return (
+          matchedIntentNode?.target &&
+          this.getNodeById<HtNodeWithContent>(matchedIntentNode?.target.id)
+        )
+      }
+    } catch (error) {
+      console.error(`Error getting node by intent '${input.intent}': `, error)
+    }
+
+    return undefined
+  }
+
+  private hasIntent(
+    node: HtIntentNode,
+    intent: string,
+    locale: string
+  ): boolean {
+    return node.content.intents.some(
+      i => i.locale === locale && i.values.includes(intent)
+    )
+  }
+
+  private hasMetConfidenceThreshold(
+    node: HtIntentNode,
+    predictedConfidence: number
+  ): boolean {
+    const nodeConfidence = node.content.confidence / 100
+    return predictedConfidence >= nodeConfidence
+  }
+
+  getNodeByKeyword(
+    userInput: string,
+    locale: string
+  ): HtNodeWithContent | undefined {
+    try {
+      const keywordNodes = this.flow.nodes.filter(
+        node => node.type == HtNodeWithContentType.KEYWORD
+      ) as HtKeywordNode[]
+      const matchedKeywordNodes = keywordNodes.filter(node =>
+        this.matchKeywords(node, userInput, locale)
+      )
+      if (matchedKeywordNodes.length > 0 && matchedKeywordNodes[0].target) {
+        return this.getNodeById<HtNodeWithContent>(
+          matchedKeywordNodes[0].target.id
+        )
+      }
+    } catch (error) {
+      console.error(`Error getting node by keyword '${userInput}': `, error)
+    }
+
+    return undefined
```

Comment:
looking good! i'll try to do early-return instead: https://szymonkrajewski.pl/why-should-you-return-early/

Reply:
This code has just been moved from one file to another, I haven't done any change to it and any change to be applied here I prefer to do it into a separated PR.
I agree with you that this code can be improved, we can do it in the next iteration :wink:!